Add initial pages + pagesQuery endpoint to /replay.json APIs

ikreymer opened the pull request:
Fixes #2360 

- Adds `seedPages` to /replay.json response for collections, returning upto 25 seed pages.
- Adds `pagesQueryUrl` to /replay.json
- Adds a public pages search endpoint to support public collections.
- Adds `preloadResources`, including list of WACZ files that should always be loaded, to /replay.json

Draft pending work in wabac.js to ensure this is complete.

ikreymer committed:
add pages to coll replay.json endpoint

ikreymer committed:
add pages to coll replay.json endpoint

ikreymer committed:
create CollOut after

ikreymer committed:
create CollOut after

ikreymer committed:
add pagesQuery to replay.json

ikreymer committed:
add pagesQuery to replay.json

ikreymer committed:
update model

ikreymer committed:
update model

ikreymer committed:
add /public/pages endpoint for public collections
tests: add test 'pagesQuery' for both private and public collections

ikreymer committed:
add /public/pages endpoint for public collections
tests: add test 'pagesQuery' for both private and public collections

ikreymer committed:
work on tests

ikreymer committed:
work on tests

ikreymer committed:
test work

ikreymer committed:
test work

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -1327,6 +1482,9 @@ class CollOut(BaseMongoModel):
     allowPublicDownload: bool = True
 
+    pages: List[PageOut] = []
+    pagesQuery: str = ""
```
I might suggest remaining this to `pagesQueryUrl`, or even better, `additionalPagesUrl`, to be a bit clearer on what it is, otherwise could read like it's some representation of the query that produced this result.

ikreymer replied:
renamed to pagesQueryUrl

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -1327,6 +1482,9 @@ class CollOut(BaseMongoModel):
 
     allowPublicDownload: bool = True
 
+    pages: List[PageOut] = []
```
Maybe `seedPages` would be more accurate? As-is this could give the wrong idea that this is all of the available pages in the collection.

ikreymer replied:
Could be featuredPages, though maybe that's something we'd add later.

ikreymer replied:
renamed to seedPages for now, perhaps we'll add a separate featurePages later, and allow switching between them, etc...

tw4l commented:
Do we also want to return seeds and the additional pages query url in the replay.json for crawls?

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -526,9 +539,17 @@ async def get_collection_search_values(self, org: Organization):
+            try:
+                await self.get_collection_raw(coll_id, public_or_unlisted_only)
+            except HTTPException:
+                return []
```
This could use a test case

ikreymer replied:
Added, should actually be throwing here, nice catch!

tw4l commented on backend/test/test_collections.py:
```diff
@@ -382,6 +382,9 @@ def test_get_collection_replay(crawler_auth_headers, default_org_id):
     assert data["dateEarliest"]
     assert data["dateLatest"]
     assert data["defaultThumbnailName"]
+    assert data["pagesQuery"].endswith(
```
Also want check that `pages`/`seedPages` are returned here and in the test below

ikreymer replied:
Added!

tw4l committed:
Add initial URL + title search to collection pages endpoint

Needs some testing and possible refinement

tw4l committed:
Add initial URL + title search to collection pages endpoint

Needs some testing and possible refinement

tw4l committed:
Add alwaysLoad to collection replay.json

Includes information for all WACZ files in collection that contain
seed pages or have no associated pages in the database.

tw4l committed:
Add alwaysLoad to collection replay.json

Includes information for all WACZ files in collection that contain
seed pages or have no associated pages in the database.

tw4l committed:
Add pylint comment

tw4l committed:
Add pylint comment

ikreymer committed:
crawlfileout: rename to itemId, use basename for 'name'

ikreymer committed:
crawlfileout: rename to itemId, use basename for 'name'

ikreymer committed:
switch to crawlId instead

ikreymer committed:
switch to crawlId instead

ikreymer committed:
handle CORS for public pages endpoint

ikreymer committed:
handle CORS for public pages endpoint

ikreymer committed:
fix typo

ikreymer committed:
fix typo

ikreymer committed:
refactor / cleanup from code review
- rename pages -> seedPages
- rename alwaysLoad -> preloadResources
- rename pagesQuery -> pagesQueryUrl
- optimize loading preloadResources as part of resource lookup
- remove seed page wacz files from preloadResources, can be computed from seedPages list
- tests: add additional tests for preloadResources, seedPages

ikreymer committed:
refactor / cleanup from code review
- rename pages -> seedPages
- rename alwaysLoad -> preloadResources
- rename pagesQuery -> pagesQueryUrl
- optimize loading preloadResources as part of resource lookup
- remove seed page wacz files from preloadResources, can be computed from seedPages list
- tests: add additional tests for preloadResources, seedPages

ikreymer committed:
fix test?

ikreymer committed:
fix test?

ikreymer committed:
ensure public pages access to private collecton is 404

ikreymer committed:
ensure public pages access to private collecton is 404

ikreymer committed:
add seedPages and pagesQueryUrl to crawl /replay.json for consistency
update tests

ikreymer committed:
add seedPages and pagesQueryUrl to crawl /replay.json for consistency
update tests

ikreymer committed:
fix typo

ikreymer committed:
fix typo

ikreymer commented:
> Do we also want to return seeds and the additional pages query url in the replay.json for crawls?

Added `pagesQueryUrl` and `seedPages` for crawls.

tw4l committed:
Add tests for collection pages search filter

ikreymer commented:
One more small change, what if instead of seedPages we went with `initialPages` and it included the first 25 pages?
The default sort order can be `{"$sort": {"isSeed": -1, "ts": 1}}`, so we still list all seeds, then all non-seeds, sorted by time. This would allow for infinite scroll to load additional pages w/o a query.

ikreymer committed:
seedPages -> initialPages
add totalPages
collection pages: default sort order seeds first, then by timestamp

ikreymer committed:
Merge branch 'replay-json-pages' of github.com:webrecorder/browsertrix-cloud into replay-json-pages
